ts/src/main/scala/ch/epfl/ts/traders/MarketMakerTrader.scala

```diff
@@ -0,0 +1,183 @@
+import scala.language.postfixOps
+import akka.actor.ActorLogging
+import akka.actor.ActorRef
+import akka.pattern.ask
+import akka.util.Timeout
+import ch.epfl.ts.data.ConfirmRegistration
+import ch.epfl.ts.data.Currency.CHF
+import ch.epfl.ts.data.Currency.Currency
+import ch.epfl.ts.data.Currency.USD
+import ch.epfl.ts.data.MarketBidOrder
+import ch.epfl.ts.data.MarketOrder
+import ch.epfl.ts.data.Order
+import ch.epfl.ts.data.Quote
+import ch.epfl.ts.data.Register
+import ch.epfl.ts.data.StrategyParameters
+import ch.epfl.ts.engine.{GetWalletFunds, WalletConfirm, FundWallet, WalletFunds, ExecutedAskOrder, AcceptedOrder, RejectedOrder}
+import ch.epfl.ts.engine.ExecutedBidOrder
+import ch.epfl.ts.engine.Wallet
+import ch.epfl.ts.data.ParameterTrait
+import ch.epfl.ts.data.CurrencyPairParameter
+import ch.epfl.ts.data.NaturalNumberParameter
+import ch.epfl.ts.data.TimeParameter
+import ch.epfl.ts.data.CoefficientParameter
+import scala.concurrent.duration.FiniteDuration
+import ch.epfl.ts.data.Currency
+import ch.epfl.ts.engine.MarketAsksEmpty
+import ch.epfl.ts.engine.MarketBidsEmpty
+import ch.epfl.ts.engine.MarketMakerNotification
+import ch.epfl.ts.data.LimitAskOrder
+import ch.epfl.ts.data.LimitBidOrder
+import ch.epfl.ts.engine.MarketMakerNotification
+import ch.epfl.ts.engine.MarketEmpty
+
+/**
+ * MarketMakerTrader companion object
+ * acts as a trader who's strategy is:
+ * if market demands cannot be fulfilled by other traders
+ *    if I have the capacity to fulfill them
+ *      -> make a limit order with my spread (charge) added
+ *    else
+ *      CRY very loud!
+ */
+object MarketMakerTrader extends TraderCompanion {
+  type ConcreteTrader = MarketMakerTrader
+  override protected val concreteTraderTag = scala.reflect.classTag[MarketMakerTrader]
+  
+  /** Interval between two random trades (in ms) */
+  val INTERVAL = "interval"
+
+  /** Initial delay before the first random trade (in ms) */
+  val INITIAL_DELAY = "initial_delay"
+
+  /** Volume of currency to trade (in currency unit) */
+  val ORDER_VOLUME = "order_volume"
+  /** Random variations on the volume (in percentage of the order volume, both above and below `ORDER_VOLUME`) */
+  val ORDER_VOLUME_VARIATION = "order_volume_variation"
+
+  /** Which currencies to trade */
+  val CURRENCY_PAIR = "currency_pair"
+  
+  /** Market maker specific spread */
+  val SPREAD = "spread"
+
+  override def strategyRequiredParameters: Map[Key, ParameterTrait] = Map(
+      INTERVAL -> TimeParameter,
+      ORDER_VOLUME -> NaturalNumberParameter,
+      CURRENCY_PAIR -> CurrencyPairParameter
+    )
+  override def optionalParameters: Map[Key, ParameterTrait] = Map(
+      INITIAL_DELAY -> TimeParameter,
+      ORDER_VOLUME_VARIATION -> CoefficientParameter,
+      SPREAD -> CoefficientParameter //TODO make usable with values > 1
+    )
+}
+
+/**
+ * broker-aware trader.
+ */
+class MarketMakerTrader(uid: Long, marketIds: List[Long], parameters: StrategyParameters)
```

Comment:
Merge with the market maker that uses a wallet (simply add a boolean flag in `optionalParameters` to indicate whether it has infinite funds or not).